### PR TITLE
[JENKINS-49573] JEP-200 compatibility

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-49573](https://issues.jenkins-ci.org/browse/JENKINS-49573)

Subsumes #23.

- [X] apply what I think is the correct fix
- [x] deal with serial compatibility of `ResultShortcut.resultsToCheck`
- [x] deal with serial compatibility in corner cases
- [x] handle test failures

@reviewbybees esp. @oleg-nenashev

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
```
java.lang.ClassCastException: net.sf.json.JSONNull cannot be cast to net.sf.json.JSONObject
at com.sonyericsson.rebuild.RebuildAction.doConfigSubmit(RebuildAction.java:286)
at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
at org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)
```

The test fails due to https://issues.jenkins-ci.org/browse/JENKINS-45373 IIUC

File: src/main/java/hudson/plugins/matrix_configuration_parameter/shortcut/ResultShortcut.java [hunk 1]
```diff
@@ -69,7 +70,7 @@ public ResultShortcut(String name, boolean exact, List<String> resultsToCheck) {
         this.name = Util.fixNull(name);
         this.exact = exact;
         this.resultsToCheck = (resultsToCheck != null)
-            ? Collections.unmodifiableList(resultsToCheck)
+            ? new ArrayList<>(resultsToCheck)
```
Comment (ikedam) [inline src/main/java/hudson/plugins/matrix_configuration_parameter/shortcut/ResultShortcut.java]: Note for me:
Collections.unmodifiableList results following xml:
```
              <shortcutList class="java.util.Arrays$ArrayList">
                <a class="hudson.plugins.matrix_configuration_parameter.shortcut.MatrixCombinationsShortcut-array">
                  <hudson.plugins.matrix__configuration__parameter.shortcut.ResultShortcut>
                    <name>Successful</name>
                    <exact>false</exact>
                    <resultsToCheck class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                      <c class="com.google.common.collect.Lists$TransformingRandomAccessList">
                        <fromList class="java.util.Arrays$ArrayList">
                          <a class="hudson.model.Result-array">
                            <hudson.model.Result>SUCCESS</hudson.model.Result>
                          </a>
                        </fromList>
                        <function class="hudson.plugins.matrix_configuration_parameter.shortcut.ResultShortcut$1"/>
                      </c>
                      <list class="com.google.common.collect.Lists$TransformingRandomAccessList" reference="../c"/>
                    </resultsToCheck>
                  </hudson.plugins.matrix__configuration__parameter.shortcut.ResultShortcut>
                  <hudson.plugins.matrix__configuration__parameter.shortcut.ResultShortcut>
                    <name>Failed</name>
                    <exact>false</exact>
                    <resultsToCheck class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                      <c class="com.google.common.collect.Lists$TransformingRandomAccessList">
                        <fromList class="java.util.Arrays$ArrayList">
                          <a class="hudson.model.Result-array">
                            <hudson.model.Result>FAILURE</hudson.model.Result>
                          </a>
                        </fromList>
                        <function class="hudson.plugins.matrix_configuration_parameter.shortcut.ResultShortcut$1"/>
                      </c>
                      <list class="com.google.common.collect.Lists$TransformingRandomAccessList" reference="../c"/>
                    </resultsToCheck>
                  </hudson.plugins.matrix__configuration__parameter.shortcut.ResultShortcut>
                  <hudson.plugins.matrix__configuration__parameter.shortcut.MatrixCombinationsShortcut_-All/>
                  <hudson.plugins.matrix__configuration__parameter.shortcut.MatrixCombinationsShortcut_-None/>
                </a>
              </shortcutList>
```

This means `unmodifiableList` preserves the original structure (`TransformingRandomAccessList`).
`ArrayList` receives iterator and create a new array, and remove the original structure.

File: Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// Build the plugin using https://github.com/jenkins-infra/pipeline-library
+buildPlugin(jenkinsVersions: [null, '2.107'])
```
Comment (ikedam) [inline Jenkinsfile]: Why tests with 2.107 succeeds even with `java.level=7`?

**Comment by ikedam** [comment]
When build.xml with `TransformingRandomAccessList` already exists on the disk, it is silently ignored with the following error on startup:
```
WARN: com.google.common.collect.Lists$TransformingRandomAccessList in file:/C:/Users/ikedam/Desktop/workspace/matrix-combinations-plugin/target/tmp/webapp/WEB-INF/lib/guava-11.0.1.jar might be dangerous, so rejecting; see https://jenkins.io/redirect/class-filter/
```

The build is loaded, and I suppose `resultsToCheck.c` is populated with `null`.
I think it's harmless and acceptable.

File: pom.xml [hunk 3]
```diff
@@ -21,51 +22,42 @@
   </scm>
 
     <properties>
-        <jenkins.version>1.532</jenkins.version> <!-- ParameterDefinition#getFormattedDescription is since 1.520 -->
-        <jenkins-test-harness.version>1.532</jenkins-test-harness.version>
-        <java.level>6</java.level>
+        <jenkins.version>1.625.3</jenkins.version>
```
Comment (ikedam) [inline pom.xml]: 1.651.3 might be better for SECURITY-170 affects much.

File: Jenkinsfile (hunk 2 above)
Comment (jglick) [reply Jenkinsfile]: You do not _have_ to compile using Java 8 syntax just to run on 2.60+.

**Comment by jglick** [comment]
> The build is loaded, and I suppose `resultsToCheck.c` is populated with `null`.
> I think it's harmless and acceptable.

I have this reproduced in a test and am studying the effects. There are still some issues, though they seem minor.

File: pom.xml (hunk 3 above)
Comment (jglick) [reply pom.xml]: Anyway I saw no 170-related test failures.

File: src/main/java/hudson/plugins/matrix_configuration_parameter/shortcut/ResultShortcut.java (hunk 1 above)
Comment (jglick) [reply src/main/java/hudson/plugins/matrix_configuration_parameter/shortcut/ResultShortcut.java]: Now visible in the `hairyCompat` test input.

**Comment by jglick** [comment]
@ikedam I think this is ready to merge, so please take another look when you get a chance.

File: src/test/java/hudson/plugins/matrix_configuration_parameter/MatrixCombinationsParameterDefinitionTest.java [hunk 4]
```diff
@@ -659,4 +673,38 @@ public void testSafeDescription() throws Exception {
...
+        MatrixProject p = j.jenkins.getItemByFullName("p", MatrixProject.class);
+        MatrixBuild b = j.assertBuildStatusSuccess(p.scheduleBuild2(0, (Cause) null, new ParametersAction(new MatrixCombinationsParameterValue("x", "", Arrays.asList("x=b,y=2")))));
+        for (MatrixCombinationsShortcut shortcut : ((MatrixCombinationsParameterDefinition) p.getProperty(ParametersDefinitionProperty.class).getParameterDefinition("x")).getShortcutList()) {
+            System.out.println(shortcut.getCombinationsData(p, b));
+        }
+        b.save();
+        System.out.println(new XmlFile(Run.XSTREAM, new File(b.getRootDir(), "build.xml")).asString());
+        p.save();
+        System.out.println(p.getConfigFile().asString());
+    }
```
Comment (ikedam) [inline src/test/java/hudson/plugins/matrix_configuration_parameter/MatrixCombinationsParameterDefinitionTest.java]: The output: https://gist.github.com/ikedam/d9ec50761d4c0f2020662f33606887aa

Comment (ikedam) [reply src/test/java/hudson/plugins/matrix_configuration_parameter/MatrixCombinationsParameterDefinitionTest.java]: What happens:
* Fail to deserialize `resultsToCheck#c` for JEP-200. It resuluts broken `resultsToCheck`.
* Somehow it causes NPE in `shourcutList#a`. It results `shortcutList#a` to be `null`.
* It results `shortcutList#toString()` fail, and `readResolve` sets a default value for `shortcutList`.

Comment (jglick) [reply src/test/java/hudson/plugins/matrix_configuration_parameter/MatrixCombinationsParameterDefinitionTest.java]: Right. Not the cleanest fix, but probably good enough.

File: pom.xml (hunk 3 above)
Comment (ikedam) [reply pom.xml]: I decided to target 1.651.3 as in #24 .
* The next release anyway updates the target version from 1.532.
* This plugin will depend on matrix-project-1.7 for compatibility tests pass SECURITY-170, and it looks natural to me to target core with SECURITY-170.
* It makes easier to see XMLs created by SECURITY-170.
    * Of course, I know it will be fixed in jenkinsci/jenkins#3312 .
